keps/1136-provisioning-request-support/README.md

```diff
@@ -170,6 +172,59 @@ type ProvisioningRequestConfigSpec struct {
 	// +listType=set
 	// +kubebuilder:validation:MaxItems=100
 	ManagedResources []corev1.ResourceName `json:"managedResources,omitempty"`
+
+	// retryStrategy defines strategy for retrying ProvisioningRequest
+	//
+	// +optional
+	RetryStrategy *ProvisioningRequestRetryStrategy `json:retryStrategy, omitempty`
+}
+
+type ProvisioningRequestRetryStrategy struct {
+	// Timestamp defines the timestamp used for re-queuing a Workload
+	// that was evicted due to Pod readiness. The possible values are:
+	//
+	// - `Eviction` (default) indicates from Workload `Evicted` condition with `PodsReadyTimeout` reason.
+	// - `Creation` indicates from Workload .metadata.creationTimestamp.
+	//
+	// +optional
+	Timestamp *RequeuingTimestamp `json:"timestamp,omitempty"`
+
+	// BackoffLimitCount defines the maximum number of re-queuing retries.
+	// Once the number is reached, the workload is deactivated (`.spec.activate`=`false`).
+	// When it is null, the workloads will repeatedly and endless re-queueing.
+	//
+	// Every backoff duration is about "b*2^(n-1)+Rand" where:
+	// - "b" represents the base set by "BackoffBaseSeconds" parameter,
+	// - "n" represents the "workloadStatus.requeueState.count",
+	// - "Rand" represents the random jitter.
+	// During this time, the workload is taken as an inadmissible and
+	// other workloads will have a chance to be admitted.
+	// By default, the consecutive requeue delays are around: (60s, 120s, 240s, ...).
+	//
+	// Defaults to null.
+	// +optional
+	BackoffLimitCount *int32 `json:"backoffLimitCount,omitempty"`
+
+	// BackoffBaseSeconds defines the base for the exponential backoff for
+	// re-queuing an evicted workload.
+	//
+	// Defaults to 60.
+	// +optional
+	BackoffBaseSeconds *int32 `json:"backoffBaseSeconds,omitempty"`
+
+	// BackoffMaxSeconds defines the maximum backoff time to re-queue an evicted workload.
+	//
+	// Defaults to 3600.
+	// +optional
+	BackoffMaxSeconds *int32 `json:"backoffMaxSeconds,omitempty"`
+
+	// releaseQuota defines whether the reserved quota should be released during retry.
+	// If the releaseQuota is set not false, controller will delete old ProvisioningRequest
+	// and create a new one without releasing quota.
+	//
+	// Defaults to true.
+	// +optional
+	Requeue bool `json:requeue`
```

Comment:
```suggestion
	ReleaseQuota bool `json:releaseQuota`
```

Which conditions (QuotaReservation vs Admitted) will we turn to false?

Reply:
We will mark `QuotaReserved` as false. The Admitted condition is not set to true at this stage as this condition is only set after all AdmissionChecks succeed

Reply:
Do you think ReleaseQuota is better naming from the user perspective? Ultimately it's not only about releasing quota but also requeuing the Workload (which involves releasing quota)

Reply:
I discussed this with @PBundyra offine.
This feature tries to introduce the requeuing Workload (WL) mechanism when the AdmissionCheck (AC) gets the `Rejected` condition due to ProvisioningRequest (PR) getting the `Failed` condition.

The current ProvisioningRequest deactivation mechanism is Failed (PR) -> Rejected (AC) -> Deactivated (WL). So, the WL will be deactivated without requeueing.
Hence, this feature will introduce the requeuing mechanism to `Rejected (AC) -> Deactivated (WL)` phase.

This allows us the possibility to distinguish between infrastructure and application configuration errors because, after this feature, the WaitForPodsReady indicates only application configuration errors due to capacity checks can be made before admission. But distinguishing between both error reasons could be considered as a separate feature from this feature. 

So, we will implement the distinguishness after investigating the values, and we will focus only on the PR requeuing feature here.

Reply:
That's right. The only thing I would clarify is that `Failed (PR)` depending on configuration can mean two things `Retry (AC)` or `Rejected (AC)`. Currently, we never set `Retry (AC)` if the ProvReq fails. Hence, the change would be to set the `Retry (AC)` condition, and then handle it accordingly to the above description. If we exhaust number of retries than the `Rejected (AC)` condition will be set, and we will follow the current behavior which is to deactivate the Workload

Reply:
Thanks for extending my description. That would be great If you can clarify the current behavior and updated behavior in this KEP.

Reply:
Done